src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketBranchClientImpl.java

```diff
@@ -0,0 +1,71 @@
+package com.atlassian.bitbucket.jenkins.internal.client;
+
+import com.atlassian.bitbucket.jenkins.internal.client.paging.BitbucketPageStreamUtil;
+import com.atlassian.bitbucket.jenkins.internal.client.paging.NextPageFetcher;
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketDefaultBranch;
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketPage;
+import com.fasterxml.jackson.core.type.TypeReference;
+import okhttp3.HttpUrl;
+
+import java.util.Collection;
+import java.util.stream.Stream;
+
+import static java.lang.String.valueOf;
+
+public class BitbucketBranchClientImpl implements BitbucketBranchClient {
+
+    private final BitbucketRequestExecutor bitbucketRequestExecutor;
+    private final String projectKey;
+    private final String repositorySlug;
+
+    public BitbucketBranchClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor,
+                                     String projectKey,
+                                     String repositorySlug) {
+        this.bitbucketRequestExecutor = bitbucketRequestExecutor;
+        this.projectKey = projectKey;
+        this.repositorySlug = repositorySlug;
+    }
+
+    @Override
+    public Stream<BitbucketDefaultBranch> getRemoteBranches() {
+        HttpUrl.Builder urlBuilder = bitbucketRequestExecutor.getCoreRestPath().newBuilder()
+                .addPathSegment("projects")
+                .addPathSegment(projectKey)
+                .addPathSegment("repos")
+                .addPathSegment(repositorySlug)
+                .addPathSegment("branches")
+                .addQueryParameter("filterTest", "");
```

Comment:
```suggestion
                .addQueryParameter("filterText", "");
```

Also, I think it's optional, so we can probably omit it.